gdal/port/cpl_json.cpp

```diff
@@ -0,0 +1,1035 @@
+ * @param  pProgressArg application data passed into progress function.
+ * @return              true on success. If error occured it can be received using CPLGetLastErrorMsg method.
+ *
+ * @since GDAL 2.3
+ */
+
+#ifdef HAVE_CURL
+bool CPLJSONDocument::LoadUrl(const char *pszUrl, char **papszOptions,
+                              GDALProgressFunc pfnProgress,
+                              void *pProgressArg)
+#else
+bool CPLJSONDocument::LoadUrl(const char * /*pszUrl*/, char ** /*papszOptions*/,
+                              GDALProgressFunc /*pfnProgress*/,
+                              void * /*pProgressArg*/)
+#endif // HAVE_CURL
+{
+#ifdef HAVE_CURL
+    int nDepth = atoi( CSLFetchNameValueDef( papszOptions, "JSON_DEPTH", "10") );
+    JsonContext ctx = { nullptr, json_tokener_new_ex(nDepth), 0 };
+    CPLHTTPResult *psResult = CPLHTTPFetchEx( pszUrl, papszOptions,
+                                              pfnProgress, pProgressArg,
+                                              CPLJSONWriteFunction, &ctx );
+
+    bool bResult = true;
+    if( psResult->nStatus != 0 /*CURLE_OK*/ )
+    {
+        bResult = false;
+    }
+
+    CPLHTTPDestroyResult( psResult );
+
+    enum json_tokener_error jerr;
+    if ((jerr = json_tokener_get_error(ctx.pTokener)) != json_tokener_success) {
+        CPLError(CE_Failure, CPLE_AppDefined, "JSON error: %s\n",
+               json_tokener_error_desc(jerr));
+        bResult = false;
+    }
+    else {
+        m_poRootJsonObject = ctx.pObject;
+    }
+    json_tokener_free(ctx.pTokener);
+
+    return bResult;
+#else
+    return false;
+#endif
+}
+
+//------------------------------------------------------------------------------
+// JSONObject
+//------------------------------------------------------------------------------
+/*! @cond Doxygen_Suppress */
+CPLJSONObject::CPLJSONObject() : m_poJsonObject(json_object_new_object())
+{
+
+}
+
+CPLJSONObject::CPLJSONObject(const char *pszName, const CPLJSONObject &oParent) :
+    m_poJsonObject(json_object_get(json_object_new_object())),
+    m_soKey(pszName)
+{
+    json_object_object_add( TO_JSONOBJ(oParent.m_poJsonObject), pszName,
+                            TO_JSONOBJ(m_poJsonObject) );
+}
+
+CPLJSONObject::CPLJSONObject(const CPLString &soName, JSONObjectH poJsonObject) :
+    m_poJsonObject( json_object_get( TO_JSONOBJ(poJsonObject) ) ),
+    m_soKey(soName)
+{
+
+}
+
+CPLJSONObject::~CPLJSONObject()
+{
+    // Should delete m_poJsonObject only if CPLJSONObject has no parent
+    json_object_put( TO_JSONOBJ(m_poJsonObject) );
+}
+
+CPLJSONObject::CPLJSONObject(const CPLJSONObject &other) :
+    m_poJsonObject(json_object_get( TO_JSONOBJ(other.m_poJsonObject) )),
+    m_soKey(other.m_soKey)
+{
+}
+
+CPLJSONObject &CPLJSONObject::operator=(const CPLJSONObject &other)
+{
+    m_soKey = other.m_soKey;
+    m_poJsonObject = json_object_get( TO_JSONOBJ(other.m_poJsonObject) );
```

Comment:
I guess a json_object_put( TO_JSONOBJ(m_poJsonObject) ) would be needed to avoid a leak ?

Reply:
Fixed